completion_aggregator/management/commands/migrate_progress.py

```diff
@@ -0,0 +1,139 @@
+            type=int,
+        )
+        parser.add_argument(
+            '--start-index',
+            help='Offset from which to start processing CourseModuleCompletions.',
+            default=0,
+            type=int,
+        )
+
+    def handle(self, *args, **options):
+        if not PROGRESS_IMPORTED:
+            print("Unable to import progress models.  Aborting")
+            sys.exit(1)
+        self._set_log_level(options)
+        task_options = self.get_task_options(options)
+        cmc_count = CourseModuleCompletion.objects.all().count()
+        for index in range(options['start_index'], cmc_count, options['batch_size']):
+            migrate_batch.apply_async(
+                kwargs={'offset': index, 'batch_size': options['batch_size']},
+                **task_options
+            )
+
+    def get_task_options(self, options):
+        """
+        Return task options for generated celery tasks.
+
+        Currently, this adds a routing key, if provided.
+        """
+        opts = {}
+        if options.get('routing_key'):
+            opts['routing_key'] = options['routing_key']
+        return opts
+
+    def _set_log_level(self, options):
+        """
+        Sets logging levels for this module and the block structure
+        cache module, based on the given the options.
+        """
+        if options.get('verbosity') == 0:
+            log_level = logging.WARNING
+        elif options.get('verbosity') >= 1:
+            log_level = logging.INFO
+        log.setLevel(log_level)
+
+
+@celery.shared_task
+def migrate_batch(offset, batch_size):  # Cannot pass a queryset to a task.
+    """
+    Convert a batch of CourseModuleCompletions to BlockCompletions.
+
+    Given an offset and batch_size, this task will:
+
+    * Fetch a subset of the existing CourseModuleCompletions,
+    * Update the BlockCompletions table
+    * Run aggregators for each one.
+    """
+
+    # Keep the ordering consistent with the index.
+    queryset = CourseModuleCompletion.objects.all().order_by('course_id', 'content_id', 'user')
+    course_module_completions = queryset[offset:offset + batch_size]
+    post_save.disconnect(completion_updated_handler, sender=BlockCompletion)
+    processed = {} # {course: {user: [blocks]}
+    for cmc in course_module_completions:
+        try:
+            course_key = CourseKey.from_string(cmc.course_id)
+            block_key = UsageKey.from_string(cmc.content_id).map_into_course(course_key)
+        except InvalidKeyError:
+            log.exception(
+                "Could not migrate CourseModuleCompletion with values: %s",
+                cmc.__dict__,
+            )
+            continue
+        if course_key not in processed:
+            processed[course_key] = {}
+        if cmc.user not in processed[course_key]:
+            processed[course_key][cmc.user] = set()
+        processed[course_key][cmc.user].add(block_key)
+        BlockCompletion.objects.submit_completion(
```

Comment:
Just confirming that this is idempotent?  If we have to stop this half way through and re-start it,  it won't make multiple entries right?

Reply:
That's correct.  submit_completion does no writes unless something has changed.  There's also a command line argument to allow specifying a start offset, so we can reduce the amount of work done in case we have to restart it.

Reply:
Given that this is distributed, is there an easy query we can run to figure out what a good offset is.  Also, given that this sends async tasks which are harder to cancel, does it make sense to have a stop index so we can do a few small initial batches before we kick the whole thing off?

Reply:
I'll address these as follows:
* Add a log statement indicating the offset and batch size of each task as it completes.  
* Add a stop-index argument.